app/uk/gov/hmrc/customs/declarations/information/services/declarationsInformationServices.scala

```diff
@@ -193,59 +187,38 @@ abstract class DeclarationService @Inject()(override val apiSubFieldsConnector:
     val dateTime = dateTimeProvider.nowUtc()
     val correlationId = uniqueIdsService.correlation
 
-    futureApiSubFieldsId(asr.clientId) flatMap {
-      case Right(sfId) =>
-        connector.send(dateTime, correlationId, asr.requestedApiVersion, sfId, searchType)
-          .map(response => {
-            val filteredResponse = filterResponse(response, XML.loadString(response.body))
-            logFilteringDuration(LocalDateTime.now)
-            Right(filteredResponse)
-          })
-          .recover(recoverException(asr))
+    futureApiSubFieldsId(asr.clientId).flatMap {
       case Left(result) =>
         Future.successful(Left(result))
+      case Right(sfId) =>
+        connector.send(dateTime, correlationId, asr.requestedApiVersion, sfId, searchType)
+          .map {
+            case Right(response) =>
+              logger.warn(s"Response is ${response.body}")
+              val filteredResponse = filterResponse(response, XML.loadString(response.body))
+              Right(filteredResponse)
+            case Left(RetryError) =>
+              logger.error("Unhealthy state entered so returning 500 to consumer with message service unavailable")
+              Left(errorResponseServiceUnavailable.XmlResult.withConversationId)
+            case Left(UnexpectedError(t)) =>
+              logger.error(s"declaration [$endpointName] call failed: [${t.getMessage}] so returning 500 to consumer", t)
+              Left(ErrorInternalServerError.XmlResult.withConversationId)
+            case Left(Non2xxResponseError(status, body)) =>
+              status match {
+                case INTERNAL_SERVER_ERROR =>
+                  val errorCodeText = (XML.loadString(body) \ "errorCode").text
+                  val errorResponse: ErrorResponse = matchErrorCode(errorCodeText)
+                  logger.warn(s"declaration [$endpointName] call failed with backend http status code of [500] and error: [${errorResponse.errorCode}] so returning to consumer " +
+                    s"[${errorResponse.httpStatusCode}] and response body: [${errorResponse.XmlResult.body.asInstanceOf[HttpEntity.Strict].data.utf8String}]")
+                  Left(errorResponse.XmlResult.withConversationId)
+                case FORBIDDEN =>
+                  logger.warn(s"declaration [$endpointName] call failed with backend http status code of [403] so returning to consumer [403]")
+                  Left(ErrorPayloadForbidden.XmlResult.withConversationId)
+                case unexpectedStatus =>
+                  logger.error(s"declaration [$endpointName] call failed with unexpected backend http status code of [$unexpectedStatus] so returning to consumer [500]")
+                  Left(ErrorInternalServerError.XmlResult.withConversationId)
+              }
+          }
```

Comment:
Is it possible to introduce a method to handleResult to reduce duplication as in the example [here] (https://github.com/hmrc/customs-inventory-linking-imports/blob/a77d15aa952200aaa0c5d9cbc59348dc4beaf430/app/uk/gov/hmrc/customs/inventorylinking/imports/services/MessageSender.scala#L91)

Reply:
Done, though it's not how I prefer to do it as it introduces an extra level of indirection just to save a couple of lines.